File: sumRoots.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def findSum(self, root, sums, cur):
        
        #base
        if root is None:
            return 0
         
        #logic
        cur = cur * 10 + root.val
        if ((root.left == None) & (root.right == None)):
            sums += cur
            return sums
        return self.findSum(root.left, sums, cur) + self.findSum(root.right, sums, cur)
      
        
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root is None:
            return 0
        
        return self.findSum(root, 0, 0)
        
        

File: sumRoots.java
// Time Complexity : O(n) 
// Space Complexity : O(h) h-height of the tree
// Did this code successfully run on Leetcode : Yes
public class sumRoots {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    int sum;
    public int sumNumbers(TreeNode root) {
        if(root == null) return sum;
        helper(root, 0);
        return sum;
    }
    
    private void helper(TreeNode root, Integer num){
        if(root == null) return;
        helper(root.left, num * 10 + root.val);
        if(root.left == null && root.right == null){
            sum += num * 10 + root.val;
        }
        helper(root.right, num * 10 + root.val);
    } 
}
